Mathlib/Topology/Compactification/OnePointEquiv.lean

```diff
@@ -0,0 +1,155 @@
+noncomputable def divSlope [DivisionRing K] (p : ℙ K (Fin 2 → K)) : OnePoint K :=
+  Quotient.lift (fun u => divOnePoint (u.1 0) (u.1 1)) divOnePoint_lifts p
+
+/-! ### Equivalence
+We establish the equivalence between `OnePoint K` and `ℙ K (Fin 2 → K)` for a division ring `K`.
+-/
+
+
+/-- In a division ring, if `a ≠ 0 ≠ b` then `(a b⁻¹) (b a⁻¹) = 1`. -/
+lemma rev_div_assoc {K : Type*} [DivisionRing K] {a b : K}
+    (ga : a ≠ 0) (gb : b ≠ 0) : a * b⁻¹ * (b * a⁻¹) = 1 :=
+  (IsUnit.mul_eq_one_iff_inv_eq (by simp_all)).mpr (by simp_all)
+
+/-- `divSlope` respects projective equivalence. -/
+lemma divSlope_inj_lifted [DivisionRing K] (a b : {v : Fin 2 → K // v ≠ 0})
+    (h : divSlope ⟦a⟧ = divSlope ⟦b⟧) : (⟦a⟧ : Quotient (projectivizationSetoid K _)) = ⟦b⟧ :=
+  Quotient.sound <| by
+  by_cases ga : a.1 1 = 0
+  · by_cases gb : b.1 1 = 0
+    · simp only [ne_eq, Decidable.not_not]
+      have : a.1 0 ≠ 0 := fun hc => .elim <|a.2 <|funext fun j => by fin_cases j <;> simp [hc,ga]
+      have : b.1 0 ≠ 0 := fun hc => .elim <|b.2 <|funext fun j => by fin_cases j <;> simp [hc,gb]
+      use Units.mk ((a.1 0) / (b.1 0)) ((b.1 0) / (a.1 0)) (by field_simp) (by field_simp)
+      apply List.ofFn_inj.mp
+      simp only [List.ofFn_succ, Pi.smul_apply, Fin.succ_zero_eq_one]
+      rw [ga, gb]
+      field_simp
+    · simp [divSlope, divOnePoint, ga, gb] at h
+  · by_cases gb : b.1 1 = 0
+    · simp [divSlope, divOnePoint, ga, gb] at h
+    · use Units.mk (a.1 1 * (b.1 1)⁻¹) (b.1 1 * (a.1 1)⁻¹)
+        (rev_div_assoc ga gb) (rev_div_assoc gb ga)
+      ext s
+      fin_cases s
+      · simp only [divSlope, divOnePoint, ite_not, Quotient.lift_mk, smul_eq_mul, ga, gb] at h
+        have h' : (a.1 1)⁻¹ * a.1 0  = (b.1 1)⁻¹ * b.1 0 := Option.some_injective K h
+        simp only [ne_eq, Fin.isValue, Fin.zero_eta, Pi.smul_apply, Units.smul_mk_apply,
+          smul_eq_mul]
+        rw [mul_assoc]
+        rw [← h']
+        rw [← mul_assoc]
+        simp only [Fin.isValue, ne_eq]
+        field_simp at h' ⊢
+      · simp [h, ga, gb]
+
+/-- Over any division ring `K`, `divSlope` is injective. -/
+lemma divSlope_injective [DivisionRing K] : Function.Injective (@divSlope K _) :=
+  Quotient.ind (fun a ↦ Quotient.ind (divSlope_inj_lifted a))
+
+/-- An inverse of `divSlope`. -/
+def slope_inv [DivisionRing K] (p : OnePoint K) : ℙ K (Fin 2 → K) := match p with
+|some t => mk' K ⟨![t, 1], by simp⟩
+|∞      => mk' K ⟨![1, 0], by simp⟩
+
+
+/-- `slope_inv` is an inverse of `divSlope`. -/
+lemma divSlope_inv [DivisionRing K] : Function.LeftInverse (@divSlope K _) slope_inv := by
+  intro a
+  have g₀       : divOnePoint' ⟨![(1:K), 0], by simp⟩ =      ∞ := by simp [divOnePoint']
+  have g₁ (t:K) : divOnePoint' ⟨![    t, 1], by simp⟩ = some t := by simp [divOnePoint']
+  cases a with
+  |none => exact g₀
+  |some t => exact g₁ t
+
+/-- `divSlope` is surjective. -/
+lemma divSlope_surjective [DivisionRing K] : Function.Surjective (@divSlope K _) :=
```

Comment:
Maybe this can be removed? I guess the point is to use the equivalence rather than the bare function.

Reply:
Good idea... done